### DISCUSSION TITLE
Fix issue575 "Nested generics seem to have issues with support."

### DISCUSSION THREAD

**Discussion by abing22333** [TongchengOpenSource/smart-doc#577] [Q&A] [answered]
### Bug Cause
The issue in [issue575](https://github.com/smart-doc-group/smart-doc/issues/575) is:

The underlying cause is that QDox fails to parse the generics of method return values in interfaces.

However, smart-doc provides support for the JPA Page interface, but the code in ParamsBuildHelper was modified, not JsonBuildHelper. Only the generated JSON data is correct.

### Test Case

The test case is in JpaController.

#### Response-fields:

```html
<body>
<!--StartFragment-->

content | array | No comments found. | -
-- | -- | -- | --
└─any object | object | any object. | -

<!--EndFragment-->
</body>
</html>
```

#### Response-example:

```json
 "content": [
    {
      "roleId": 0,
      "roleName": "",
      "age": 0
    }
  ]
```
### JsonBuildHelper Code
![image](https://github.com/smart-doc-group/smart-doc/assets/50514081/31c2f85c-1001-4f95-94f7-a1947de73eb0)

### ParamsBuildHelper Code
![image](https://github.com/smart-doc-group/smart-doc/assets/50514081/70cea857-4f11-466b-9f3d-acc2034ce002)

### Commit History

<html>
<body>
8b1aa8b639028bd57f5f1e667999a31fc33b279e 
</body>
</html>

![image](https://github.com/smart-doc-group/smart-doc/assets/50514081/80bc0844-d5fb-4e5c-a290-54b12f745fb0)

### Questions

I cannot understand the scenario of commit 8b1aa8b6, and after reviewing the change logs from May and June 22, there is no relevant information.

Can we directly revert this commit?

**Comment by abing22333** [comment]
@chenqi146 麻烦有空时解答一下

**Reply by abing22333** [reply]
[ issues 575 ](https://github.com/smart-doc-group/smart-doc/issues/575 ) 的问题是没有加载到源码，导致QDox解析不到泛型。但是ParamsBuildHelper和 JsonBuildHelper中这两处的代码不一致，不会有问题吗？

**Reply by shalousun** [reply]
等我我有时间我来看下

**Reply by shalousun** [reply]
2.7.3开始smart-doc新增了JavaClassUtil.getFieldGenericType通过反射处理对象的范型字段。但是遗漏了对接口中范型方法的处理。因此可以新增一个处理范型方法的补偿。测试代码如下：
```
 public static void main(String[] args) {
        try {
            Class c = Class.forName("org.springframework.data.domain.Slice");
            Method[] methods = c.getDeclaredMethods();
            // c.getDeclaredMethod() 修改smart-doc时可以通过这个获取到方法名直接翻着
            for (Method m:methods) {
                
                System.out.println("methodName:"+m.getName());
                System.out.println("modifiers:"+m.getModifiers());
                //获取到范型的方法名
                Type t = m.getGenericReturnType();
                System.out.println("method type:"+t.getTypeName());
            }
        } catch (ClassNotFoundException e) {
           e.fillInStackTrace();
        }
}
```
通过上面这个测试代码参考JavaClassUtil.getFieldGenericType补充对接口方法的获取即可

**Reply by chenqi146** [reply]
抱歉, 现在才看到消息, 无太大印象了, 我查一下我代码

**Reply by chenqi146** [reply]
@abing22333 
### 在2.4.5版本下
* fieldGicName字段不携带泛型, 值为"java.util.List"
* 此行代码会被覆盖到, 主要解决未读取到源码情况下, object被替换成上一层对象展示出来, 修改后调整为显示any object
![@H0M8R Q$HI5`1S~@5}D(ZQ](https://github.com/TongchengOpenSource/smart-doc/assets/36069431/a67bfebf-c75a-447b-b090-1f6045524dbc)


![UF$ANKV(1DYL%MN~ }`NBXE](https://github.com/TongchengOpenSource/smart-doc/assets/36069431/e4fe98f3-1511-4ce7-988a-6f3aeaf8df7a)

### 而在当前2.7.8版本下
* fieldGicName字段值变化了, 携带泛型信息了, 值为"java.util.List<T>"
* 此行代码回退到之前版本

![I BGY0% J25B)O@`2@$ X3E](https://github.com/TongchengOpenSource/smart-doc/assets/36069431/e2acf26f-0fda-4934-8bce-abbd59a9440e)

**Comment by shalousun** [comment, marked as answer]
fix by #587
